### PR TITLE
Created model and migration files for behaviour and behaviour level details

### PR DESCRIPTION
**Pull request by vips11**
## Notion ticket link
[Ticket Name](https://www.notion.so/uwblueprintexecs/Extend-Behaviours-Entity-I-10d10f3fb1dc80988716e2865f807120?pvs=4)

## Implementation description
* Added a model for `behaviour_level_details` and a migration file with up and down functions
* Updated the `behaviour` model with `parent_behaviour_id` column and created a new migration file

## Steps to test
1. Run `node migration up`
2. The new table `behaviour_level_details` should be created
3. Run `node migrate down` twice
4. The new column(`parent_behaviour_id`) from `behaviours` should be removed and `behaviour_level_details` should be removed.

## What should reviewers focus on?
* The type for level is `INTEGER` instead of `ENUM` since enums cannot have integer values. The validation is instead done in the migration file for `behaviour_level_details` to check that the min value is 1 and max value is 4.

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/migrations/2024.10.22T19.38.07.create-behaviour-level-details-table.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+export const up: Migration = async ({ context: sequelize }) => {
+  await sequelize.getQueryInterface().createTable(TABLE_NAME, {
+    id: {
+      type: DataType.INTEGER,
+      allowNull: false,
+      primaryKey: true,
+      autoIncrement: true,
+    },
+    behaviour_id: {
```
Comment (jerry-cheng5) [inline backend/typescript/migrations/2024.10.22T19.38.07.create-behaviour-level-details-table.ts:15]: there should be a foreign key here to the behaviours table

File: backend/typescript/migrations/2024.10.22T19.38.07.create-behaviour-level-details-table.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  await sequelize.getQueryInterface().addConstraint(TABLE_NAME, {
+    fields: ["behaviour_id", "level"],
+    type: "unique",
+    name: "unique_behaviour_level",
+  });
+};
+
+export const down: Migration = async ({ context: sequelize }) => {
+  await sequelize.getQueryInterface().dropTable(TABLE_NAME);
```
Comment (jerry-cheng5) [inline backend/typescript/migrations/2024.10.22T19.38.07.create-behaviour-level-details-table.ts:41]: I think you might need to drop the constraint too? I know things can get messy with the ordering of which you drop the constraints and the tables it is associated with.

You can play around with this, but the general success criteria is that you want to be able to run the up script and then the down script for your migration file, and the database should stay exactly the same

File: backend/typescript/migrations/2024.10.22T19.42.10.add-behaviour-columns.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { DataType } from "sequelize-typescript";
+import { Migration } from "../umzug";
+
+export const up: Migration = async ({ context: sequelize }) => {
+  await sequelize
+    .getQueryInterface()
+    .addColumn("behaviours", "parent_behaviour_id", {
+      type: DataType.INTEGER,
```
Comment (jerry-cheng5) [inline backend/typescript/migrations/2024.10.22T19.42.10.add-behaviour-columns.ts:8]: you should have a foreign key here to itself. something like:

    pet_id: {
      type: DataType.INTEGER,
      allowNull: false,
      references: {
        model: "pets",
        key: "id",
      },
    },

(taken from backend/typescript/migrations/2024.07.06T15.03.01.create-pet-behaviour-table.ts)

File: backend/typescript/models/behvaiourLevelDetails.model.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  @ForeignKey(() => Behaviour)
+  @Column({ type: DataType.INTEGER, allowNull: false })
+  behaviour_id!: number;
+
+  @Column({ type: DataType.ENUM, allowNull: false })
+  level!: BehaviourLevel;
+
+  @Column({ type: DataType.STRING, allowNull: false })
+  description!: string;
```
Comment (jerry-cheng5) [inline backend/typescript/models/behvaiourLevelDetails.model.ts:24]: I think we can let description and training_instructions to be nullable.

Please change this here in the model, and also the migration file. (sorry I didn't specify this in the ticket)

**Review by jerry-cheng5 — CHANGES_REQUESTED** [review]
Great work! Good job for figuring out the constraint stuff, I think that's the first one for this project :)

Just a few minor details to touch up on before merging. Lmk if you have any questions!

File: backend/typescript/models/behvaiourLevelDetails.model.ts
Comment (MinIcecream) [inline backend/typescript/models/behvaiourLevelDetails.model.ts:1]: File name has a typo

File: backend/typescript/models/behvaiourLevelDetails.model.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  Column,
+  Model,
+  Table,
+  ForeignKey,
+  DataType,
+} from "sequelize-typescript";
+import Behaviour from "./behaviour.model";
+import { BehaviourLevel } from "../types";
+
+@Table({ timestamps: false, tableName: "behvaiour_level_details" })
```
Comment (MinIcecream) [inline backend/typescript/models/behvaiourLevelDetails.model.ts:11]: Also misspelled here

**Review by MinIcecream — COMMENTED** [review]
Yeah mostly looks good, I agree w what Jerry said

**Review by jerry-cheng5 — APPROVED** [review]
Looks good! I tested it by migrating all the way down then all they way back up and it works :) 

The fields look right too!

Good work Vipasha!

**Review by MinIcecream — APPROVED** [review]
Looks good!

**Pull request merged**